File: scripts/mpd_control.py
from mpd import MPDClient
import json, random, sys, time

# This script directly controls the playback of the MPD.
# It accepts CLI arguments /path/to/json, mpd-socket, and mpd-control-port

if len(sys.argv) != 3:
    print("Usage: python mpd_control.py /path/of/json [mpd socket] [mpd control port]")
    sys.exit(1)

radio = open(sys.argv[1])
mpd_socket = (sys.argv[2])
# mpd_control_port = int(sys.argv[3])

client=MPDClient()
client.connect("/run/mpd/"+mpd_socket)

# import json file as python dictionary
data=json.load(radio)
all_stations=["bouncefm", "csr", "kdst", "kjah", "krose", "mastersounds", "playbackfm", "radiols", "radiox", "sfur"]

# some stations do not have certain types of audio tracks
match mpd_socket:
    case "kjah":
        categories = ["Songs", "DJ", "ID"]
        chances = [1, 1, 1]
    case "radiols":
        categories = ["Songs", "DJ", "ID"]
        chances = [1, 1, 1]
    case "sfur":
        categories = ["Songs", "Caller", "DJ"]
        chances = [1, 0.0625, 1]
    case _:
        categories = ["Songs", "Caller", "DJ", "ID"]
        chances = [1, 0.0625, 1, 1]

for cat in categories:
    random.shuffle(data[cat])

# Define tracker variables
song_has_not_been_played_for=0
last_category=""


def main():
    # Enable consume. This allows songs to be removed from the queue after they are played. Prevents from accumulating backlog of songs
    client.consume(1)
    print("MPD Consume enabled for " + mpd_socket)
    
    while True:
        # Get song queue to be played.
        output=queue()
        for elem in output:
            #Add songs to queue one by one
            full_path=("/music/"+elem)
            print("Adding "+full_path+" to the queue.")
            client.add(full_path)
        
        sleep_for=(remaining_time())
        print("Pressing play.")
        client.play()
        
        #Sleep for the duration of the queue (not including the first track in the queue)
        print("Sleeping for "+str(sleep_for)+" seconds.")
        time.sleep(sleep_for)

def remaining_time():

    #Returns remaining duration of the queue (not including the first track in the queue)
    
    full_queue=client.playlistinfo()
    total_playtime=0
    for track in full_queue[1:]:
        total_playtime+=float(track["duration"])
    return total_playtime

def queue():
    global song_has_not_been_played_for
    global last_category
    while True:

        # Make sure a song is played at least every 3rd time
        if (song_has_not_been_played_for>1):
            selected_category="Songs"
        else:                
            # select a random category, (the list indicates the prob. weights for each category) (random.choices returns a list)
            selected_category = random.choices(categories, chances)
            # convert the list to string
            selected_category = selected_category[0]
        if (last_category):
            if(selected_category==last_category):
                continue
        
        current_track=get_next_track(selected_category)
        
        last_category=(selected_category)
        # song objects have a different structure
        # each song in radio contains at most 3 intros and 3 outros (2 with dj voice lines, 1 without)
        # each song object is a dict containing the following keys
        # title, intros (list), mid (middle part of the song, 1 element), outros (list)
        # intros (and outros) is a list containing a single dict of the following keys
        # dj0:path, dj1:path, dj2:path
        if (selected_category=="Songs"):

            # get (dict==>intros==>0th==>values) conv. to list of all possible intro lines
            # randomly choose a path from the list, assign it to song_intro
            song_intro=random.choice(list(current_track["intros"][0].values()))
            song_mid=current_track["mid"][0]["mid"]

            # same as intros but for outro
            song_outro=random.choice(list(current_track["outros"][0].values()))
            # add paths to output list
            song_has_not_been_played_for=0
            return ([song_intro, song_mid, song_outro])
        else:
            song_has_not_been_played_for+=1
            return ([current_track])
        
def get_next_track(category):
    #get the list of tracks with inside the specified category
    tracks_in_category=data[category]

    # the following lines choose 1st or 2nd track in the list randomly.
    # The chosen track is appended to the back of the list
    # This is to prevent tracks from repeating too soon, while preventing the order of the tracks from repeating as well.
    random_bit=random.randint(0,1)
    chosen_track=(tracks_in_category.pop(random_bit))
    tracks_in_category.append(chosen_track)
    return chosen_track

main()

File: scripts/mpd-control.js
// import libraries
const MPC = require("mpc-js").MPC;
const Chance = require("chance");
const chance = new Chance();
const weatherCodeReference = require("./weather.json")
// This script directly controls the playback of MPD.

// Local sockets are defined by the name of their respective stations
const channelAlias = process.env.CHANNEL_ALIAS;

// import the JSON path specified as a CLI argument
const tracksMetadata = require(`/radiosa/metadata/${channelAlias}.json`);
const stationMetadata = (require("./all-stations.json"))[channelAlias]

let categories, weights;
let songHasNotBeenPlayedFor = 0;
let lastCategory = "";

// Some stations contain certain kinds of audio tracks, but not others. 
// This switch ensures the elements are selected properly depending on the station
switch (channelAlias) {
    case "kjah":
        categories = ["Song", "DJ", "ID", "Weather", "Time of Day"];
        weights = [1, 1, 1, 0.05, 0.04];
        break;
    case "radiols":
        categories = ["Song", "DJ", "ID", "Weather", "Time of Day"];
        weights = [1, 1, 1, 0.05, 0.04];
        break
    case "sfur":
        categories = ["Song", "Caller", "DJ", "Time of Day"];
        weights = [1, 0.0625, 1, 0.04];
        break
    default:
        categories = ["Song", "Caller", "DJ", "ID", "Weather", "Time of Day"];
        weights = [1, 0.0625, 1, 1, 0.05, 0.04];
}

for (cat of categories) {
    if(!(["Weather", "Time of Day"].includes(cat))){
        tracksMetadata[cat]=chance.shuffle(tracksMetadata[cat]);
    }
}


/* 
main function calls all other functions, gets theri array output and adds them to the player queue.
then it waits for a variable amount of time before adding another track.
*/
async function main(){
    const client = new MPC();
    client.connectUnixSocket(`/radiosa/socks/${channelAlias}`);
    console.log(`Connected to MPD socket: ${stationMetadata.channel_name}.`);
    // Enable consume. This allows tracks to be removed from the player queue after they are played. Prevents from accumulating backlog of tracks
    await client.playbackOptions.setConsume(true);
    console.log(`MPD Consume enabled for ${stationMetadata.channel_name}.`);

    while (true){
        output= await getNextCategory();
        for (elem of output){
            // add tracks to queue one by one
            const fullPath=(`/radiosa/music/${elem}`)
            console.log(`Adding ${fullPath} to the queue.`)
            await client.currentPlaylist.add(fullPath)
        }

        const sleepFor = await remainingTime(client);
        await client.playback.play();

        console.log(`Sleeping for ${sleepFor} seconds.`);
        await new Promise(resolve => setTimeout(resolve, sleepFor * 1000));
    }
}

async function remainingTime(client){
    // Returns remaining duration of the queue (not including the first track in the queue)

    let fullQueue = await client.currentPlaylist.playlistInfo();
    let totalPlaytime = 0;
    for (let track of fullQueue.slice(1)){
        totalPlaytime += track.duration;
    }
    return totalPlaytime;
}


/*
* getNextCategory() decides which category of tracks is played next
* Typically, these are "Song", "Caller", "DJ", "ID", "Weather", "Time of Day".
*/
async function getNextCategory(){
    while (true) {
        let selectedCategory;
        // if 2 non-song tracks have been played, it ensures that next one is a song
        if (songHasNotBeenPlayedFor > 1 || lastCategory === "Caller"){
            selectedCategory = "Song";
        }
        else {
            selectedCategory = chance.weighted(categories, weights);
        }

        bothAreVoicedbytheDJ = () => {
            voicedbytheDJ = ("Caller", "DJ", "Weather", "Time of Day")
            return (voicedbytheDJ.includes(lastCategory) && voicedbytheDJ.includes(selectedCategory));
        }

        // ensure that 2 of the same kind of tracks are not played back to back
        if(((lastCategory) && (selectedCategory === lastCategory)) || bothAreVoicedbytheDJ()){
            continue;
        }

        let currentTrack = await getNextTrack(selectedCategory);
        if (currentTrack === null){
            continue;
        }
        lastCategory=(selectedCategory);

        /*
        song objects have a different structure than all other tracks
        each song in radio contains at most 3 intros and 3 outros (2 with dj voice lines, 1 without)
        each song object is a json object containing the following keys
        title, intros (list), mid (middle part of the song, 1 element), outros (list)
        intros (and outros) is a list containing a single dict of the following keys
        dj0:path, dj1:path, dj2:path
        */

        if (selectedCategory === "Song"){
            // get (dict==>intros==>0th==>values), convert to array of all possible intro lines
            // randomly choose a path from the list, assign it to song_intro
            
            const songIntro = chance.pickone(currentTrack.intros);
            const songMid = currentTrack.mid;
            const songOutro = chance.pickone(currentTrack.outros);

            songHasNotBeenPlayedFor = 0;
            return [songIntro, songMid, songOutro];
        }
        else {
            songHasNotBeenPlayedFor++;
            return [currentTrack];
        }
    }
}

// Provides the next track to play depending on the category that was passed
async function getNextTrack(selectedCategory){

    // weather and time of day announcements require real world data, so they are split into different functions.
    if (selectedCategory === "Weather"){
        return await getNextWeather();
    }
    else if (selectedCategory === "Time of Day"){
        return await getTimeTrack(getPacificTime);
    }
    else {
        // After receiving an array of tracks, randomly choose between indexes 0 and 1, then append that track at the back of the array
        // This FiFo like approach ensures that songs are spread apart from each other, but still random to simulate radio-like quality
        const tracksInCurrentCategory = tracksMetadata[selectedCategory];
        const randomBit = chance.bool();
        const [selectedTrack] = tracksInCurrentCategory.splice(+randomBit, 1);
        tracksInCurrentCategory.push(selectedTrack);
        return selectedTrack;
    }
}

async function getNextWeather(){
    // get current weather conditions where radio station is located
    // the location is based on an estimation of where radio stations are located in game and what that location corresponds to in real life.
    try {
        currentWeather = await fetch(`https://api.open-meteo.com/v1/forecast?latitude=${stationMetadata.latitude}&longitude=${stationMetadata.longitude}&current=weather_code`).then((response) => response.json());
    }
    catch (error){
        return null;
    }
    timeOfDay = await getPacificTime();
    // only return track if the time is morning or afternoon, so radio hosts don't say "sunny" at night or something
    isMorningorAfternoon = (timeOfDay === "Morning" || timeOfDay === "Afternoon");
    if ((weatherCodeReference[currentWeather.current.weather_code] === undefined) || !isMorningorAfternoon){
        return null;
    }
    else {
        const weatherTrack = chance.pickone(tracksMetadata.Weather[weatherCodeReference[currentWeather.current.weather_code]])
        return weatherTrack;
    }
}

async function getPacificTime(){
    // get current time of day in Pacific Timezone
    // fetch details from worldtimeapi
    const currentPacificTime = await fetch("https://worldtimeapi.org/api/timezone/America/Los_Angeles").then((response) => response.json());
    /* 
    calculate current fraction of the day by adding the offset (negative number) to the Unix Timestamp (Greenwich Time)
    Then calculate daylight savings time ONLY IF daylight savings is true
    get modulo 86400 for amount of seconds after midnight of current day
    */
    const currentTimeOfDay = (currentPacificTime.unixtime + currentPacificTime.raw_offset + (currentPacificTime.dst && currentPacificTime.dst_offset))%86400;

    if (currentTimeOfDay < 10800 || currentTimeOfDay >= 75600) return "Night";
    if (currentTimeOfDay < 43200) return "Morning";
    if (currentTimeOfDay < 54000) return "Afternoon";
    return "Evening";
}

// returns a track based on the time provided by getPacificTime function
async function getTimeTrack(getPacificTime){
    timeOfDay = await getPacificTime();

    if(tracksMetadata["Time of Day"][timeOfDay] === undefined){
        return null;
    }
    else{
        return (chance.pickone(tracksMetadata["Time of Day"][timeOfDay]));
    }
}
main();